File: sample_source_code/python/src/animal.py
"""Abstract base class for all animals."""
from __future__ import annotations

from ianimal import IAnimal


class Animal(IAnimal):
    """Provides shared state and a default describe() implementation."""

    def __init__(self, name: str, age: int, species: str, weight: float) -> None:
        self._name    = name
        self._age     = age
        self._species = species
        self._weight  = weight

    @property
    def name(self) -> str:    return self._name

    @property
    def species(self) -> str: return self._species

    @property
    def age(self) -> int:     return self._age

    @property
    def weight(self) -> float: return self._weight

    def describe(self) -> str:
        return (
            f"{self._name} ({self._species}), age={self._age}, "
            f"weight={self._weight:.1f}kg, sound='{self.sound()}'"
        )

    def __repr__(self) -> str:
        return self.describe()

File: sample_source_code/csharp/src/Animal.cs
namespace Embold.Sample;

/// <summary>
/// Abstract base class implementing IAnimal.
/// Provides shared state; subclasses override <see cref="Sound"/> and <see cref="Accept"/>.
/// </summary>
public abstract class Animal : IAnimal
{
    protected Animal(string name, int age, string species, double weight)
    {
        Name    = name;
        Age     = age;
        Species = species;
        Weight  = weight;
    }

    public string Name    { get; }
    public string Species { get; }
    public int    Age     { get; }
    public double Weight  { get; }

    public abstract string Sound();
    public abstract void   Accept(IVisitor visitor);

    public virtual string Describe() =>
        $"{Name} ({Species}), age={Age}, weight={Weight:F1}kg, sound='{Sound()}'";

    public override string ToString() => Describe();
}
